### PR TITLE
basepage and tests

### PR DESCRIPTION
**Pull request by vladskakun**

### CODE REVIEW

File: pages/basePage.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from selenium import webdriver
```
Comment (TanyaKalistruk) [inline pages/basePage.py:1]: Any docstrings?)

**Comment by TanyaKalistruk** [comment]
Good job!

File: pages/basePage.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support.wait import WebDriverWait
+from telnetlib import EC
+
+
+class BasePage:
```
Comment (TanyaKalistruk) [inline pages/basePage.py:7]: And do we have any methods for send_keys?

File: pages/basePage.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    def __init__(self):
+        """
+            Method for class fields declaration.
+        """
+        self.driver = webdriver.Chrome()
+        self.base_url = "https://eventsexpress-test.azurewebsites.net/"
+
+    def find_element_by_css(self, locator, time=10):
+        """
+            Method for search element for css selector with wait
```
Comment (TanyaKalistruk) [inline pages/basePage.py:17]: by css selector

File: pages/basePage.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        """
+        WebDriverWait(self.driver, time).until(EC.presence_of_element_located(locator),
+                                               message=f"Can't find element by locator {locator}")
+        return self.driver.find_element(By.CSS_SELECTOR, locator)
+
+    def find_element_by_xpath(self, locator, time=10):
+        """
+            Method for search element for xpath selector with wait
```
Comment (TanyaKalistruk) [inline pages/basePage.py:25]: by xpath selector

File: pages/basePage.py (hunk 3 above)
Comment (vladskakun) [reply pages/basePage.py]: done

File: pages/basePage.py (hunk 4 above)
Comment (vladskakun) [reply pages/basePage.py]: done

File: pages/Issues.py [hunk 5]
```diff
@@ -1,37 +1,40 @@
...
-        self.status_btns = ButtonElements(self.BTN_CSS)
-        self.date_from_dtp = DatePicker(self.DATAPICKER_FROM_CSS)
-        self.date_to_dtp = DatePicker(self.DATAPICKER_TO_CSS)
+    def __init__(self, driver):
+        super().__init__(driver)
+        self.status_btns = ButtonElements(self.BTN_CSS, driver)
+        self.date_from_dtp = DatePicker(self.DATAPICKER_FROM_CSS, driver)
+        self.date_to_dtp = DatePicker(self.DATAPICKER_TO_CSS, driver)
 
     def click_view(self, view_number):
```
Comment (nvoloshchuk19) [inline pages/Issues.py:22]: please, add param `view_number` to the docstring
applicable for all methods with parameters

File: pages/Issues.py [hunk 6]
```diff
@@ -1,37 +1,40 @@
...
             Method for click on issue status checkbox depending on text value.
         """
-        elements = self.driver.find_elements(By.CSS_SELECTOR, self.CHECK_CSS)
+        elements = self.find_elements(self.CHECK_CSS)
         for element in elements:
             if issue_status in element.text:
                 element.click()
                 return
+
+    def get_amount_of_issue_results(self):
```
Comment (nvoloshchuk19) [inline pages/Issues.py:39]: docstring

File: pages/Issues.py [hunk 7]
```diff
@@ -1,37 +1,40 @@
...
+        return self.find_elements(self.ISSUE_RESULTS_CSS)
```
Comment (nvoloshchuk19) [inline pages/Issues.py:40]: judging by the name here should be `return len(self.find_elements(self.ISSUE_RESULTS_CSS))`

File: pages/elements/ButtonElement.py [hunk 8]
```diff
@@ -1,30 +1,30 @@
...
 from selenium.webdriver.common.by import By
 from selenium.webdriver.support.wait import WebDriverWait
-from telnetlib import EC
+from selenium.webdriver.support import expected_conditions as EC
 
 
-class ButtonElement:
+class ButtonElement(BaseContext):
     """
         Class for click on Button by css selector.
```
Comment (nvoloshchuk19) [inline pages/elements/ButtonElement.py:10]: But `click_btn_by_xpath` method is present as well

File: pages/elements/ButtonElement.py [hunk 9]
```diff
@@ -1,30 +1,30 @@
...
+from pages.common.baseContext import BaseContext
 from selenium.webdriver import ActionChains
 from selenium.webdriver.common.by import By
 from selenium.webdriver.support.wait import WebDriverWait
-from telnetlib import EC
+from selenium.webdriver.support import expected_conditions as EC
 
 
-class ButtonElement:
+class ButtonElement(BaseContext):
```
Comment (nvoloshchuk19) [inline pages/elements/ButtonElement.py:8]: @TanyaKalistruk Do you think it's a good approach?

File: pages/common/baseContext.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+                                               message=f"Can't find element by locator {locator}")
+        return self.driver.find_element(By.XPATH, locator)
+
+    def find_elements(self, locator, time=10):
+        """
+            Method for search elements by css selector with wait
+        """
+        WebDriverWait(self.driver, time).until(EC.presence_of_all_elements_located((By.CSS_SELECTOR, locator)),
+                                               message=f"Can't find elements by locator {locator}")
+        return self.driver.find_elements(By.CSS_SELECTOR, locator)
```
Comment (nvoloshchuk19) [inline pages/common/baseContext.py:35]: empty line

File: pages/elements/ButtonElement.py [hunk 11]
```diff
@@ -1,30 +1,30 @@
...
 
     def click_btn_by_xpath(self):
         """
             Method for click on a needed button by xpath selector.
         """
-        button = self.driver.find_element(By.XPATH, self.selector)
+        button = self.find_element_by_xpath(self.selector)
         button.click()
 
     def click_by_css_with_wait(self, wait_time=10):
```
Comment (nvoloshchuk19) [inline pages/elements/ButtonElement.py:30]: I think that wait should be used by default i.e. you can move wait to `click_btn_by_css` and `click_btn_by_xpath` methods and delete this one

Comment (nvoloshchuk19) [reply pages/elements/ButtonElement.py:30]: oh, I see it's some special method with action chains, please describe it in docstring more detailed

File: tests/test_issue.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+from pages.Issues import IssuesPage
+from pages.Landing import LandingPage
+from pages.ModalPage import ModalPage
+from pages.NavigationPage import NavigationPage
+
+
+class TestsRefactor:
```
Comment (nvoloshchuk19) [inline tests/test_issue.py:7]: Don't need it in pytest

File: tests/test_issue.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    """
+        Test class
+    """
+    def test_issue_filter(self, browser):
+        self.landing = LandingPage(browser)
```
Comment (nvoloshchuk19) [inline tests/test_issue.py:12]: It should be out of the test

File: tests/test_issue.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+from pages.Landing import LandingPage
+from pages.ModalPage import ModalPage
+from pages.NavigationPage import NavigationPage
+
+
+class TestsRefactor:
+    """
+        Test class
+    """
+    def test_issue_filter(self, browser):
```
Comment (nvoloshchuk19) [inline tests/test_issue.py:11]: docstring

File: tests/test_issue.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+        self.modal.send_password_input("1qaz1qaz")
+        self.modal.click_button("Sign In")
+        self.landing.find_event_btn.click_btn_by_css()
+        self.navigation = NavigationPage(browser)
+        self.navigation.go_to_page("Issues")
+        self.issues = IssuesPage(browser)
+        self.issues.date_from_dtp.write_date_to_datepicker(21, 6, 2021)
+        self.issues.date_to_dtp.write_date_to_datepicker(15, 9, 2022)
+        self.issues.status_btns.click_btn_by_name("SEARCH")
+        assert len(self.issues.get_amount_of_issue_results()) == 2, "amount of issue results doesn`t same as expected"
```
Comment (nvoloshchuk19) [inline tests/test_issue.py:26]: expected, actual 
i.e. `2 ==  assert len(self.issues.get_amount_of_issue_results()), "amount of issue results doesnt same as expected"`

File: pages/elements/ButtonElement.py (hunk 9 above)
Comment (TanyaKalistruk) [reply pages/elements/ButtonElement.py:8]: @nvoloshchuk19 actually, I don't think that it is the best approach, but it was the only one that I could think about in three minutes for making the test runnable
Don't judge :smile:

File: tests/test_issue.py (hunk 15 above)
Comment (TanyaKalistruk) [reply tests/test_issue.py:26]: pylint doesn't like this if 2 is just a number, not a constant

File: pages/Issues.py (hunk 5 above)
Comment (vladskakun) [reply pages/Issues.py:22]: done

File: pages/Issues.py (hunk 6 above)
Comment (vladskakun) [reply pages/Issues.py:41]: done

File: tests/test_issue.py (hunk 15 above)
Comment (vladskakun) [reply tests/test_issue.py]: I get no error from pylint when I using self.issues.get_amount_of_issue_results() == 2, "amount of issue results doesn`t same as expected"

File: tests/test_issue.py (hunk 14 above)
Comment (vladskakun) [reply tests/test_issue.py:11]: done

File: pages/elements/ButtonElement.py (hunk 11 above)
Comment (vladskakun) [reply pages/elements/ButtonElement.py:30]: done

File: pages/common/baseContext.py (hunk 10 above)
Comment (vladskakun) [reply pages/common/baseContext.py]: done

File: pages/elements/ButtonElement.py (hunk 8 above)
Comment (vladskakun) [reply pages/elements/ButtonElement.py]: done

File: pages/Issues.py (hunk 7 above)
Comment (vladskakun) [reply pages/Issues.py]: done

File: tests/test_issue.py (hunk 15 above)
Comment (vladskakun) [reply tests/test_issue.py]: I must change it?

File: pages/Issues.py (hunk 5 above)
Comment (nvoloshchuk19) [reply pages/Issues.py:22]: try such an approach:
`:param viev_number: explanation`

File: pages/elements/ButtonElement.py (hunk 11 above)
Comment (nvoloshchuk19) [reply pages/elements/ButtonElement.py:30]: and then maybe worth rename the method. something like `hover_and_click_by_css`?

**Comment by nvoloshchuk19** [comment]
Also, you need to revise your requirements.txt file. There should be added pytest and selenium at least

File: tests/test_issue.py (hunk 15 above)
Comment (nvoloshchuk19) [reply tests/test_issue.py]: I voting for the adding constant or just variable `expected_result`  and then `expected_result == assert len(self.issues.get_amount_of_issue_results()), "amount of issue results doesnt same as expected"`
but the last word by Tetiana

File: pages/Issues.py (hunk 5 above)
Comment (vladskakun) [reply pages/Issues.py:22]: view_number: "type: integer. It's a parameter to select needed view button."
right?

Comment (nvoloshchuk19) [reply pages/Issues.py:22]: https://github.com/mehalyna/Ch123_TAQC_Python/pull/11/files#diff-93a0d2d69803adb31e9f9b48e52d03a8804fde6191e07f3f0ad9699ca3827eceR45

Comment (vladskakun) [reply pages/Issues.py:24]: Thank you, done.

File: pages/elements/ButtonElement.py (hunk 11 above)
Comment (vladskakun) [reply pages/elements/ButtonElement.py]: yep, done

File: tests/test_issue.py (hunk 15 above)
Comment (TanyaKalistruk) [reply tests/test_issue.py]: @vladskakun I agree with @nvoloshchuk19 last comment

File: pages/Issues.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        """
+        elements = self.find_elements(self.CHECK_CSS)
+        for element in elements:
+            if issue_status in element.text:
+                element.click()
+                return
+
+    def get_amount_of_issue_results(self):
+        """
+            Method for get amount of issue.
```
Comment (TanyaKalistruk) [inline pages/Issues.py:46]: issues

File: pages/NavigationPage.py [hunk 17]
```diff
@@ -1,20 +1,19 @@
...
     NAV_LOGOUT_CSS = "a:nth-child(2) > button"
 
-    def __init__(self):
-        self.driver = webdriver.Chrome()
-        self.navigation_edit_profile_btn = ButtonElement(self.NAV_EDIT_PROFILE_CSS)
-        self.log_out_btn = ButtonElement(self.NAV_LOGOUT_CSS)
+    def __init__(self, driver):
+        super().__init__(driver)
+        self.navigation_edit_profile_btn = ButtonElement(self.NAV_EDIT_PROFILE_CSS, self)
+        self.log_out_btn = ButtonElement(self.NAV_LOGOUT_CSS, self)
```
Comment (TanyaKalistruk) [inline pages/NavigationPage.py:16]: What is self here?

Comment (vladskakun) [reply pages/NavigationPage.py:16]: it was temporary solution. i will change it in next commit

File: tests/test_issue.py (hunk 12 above)
Comment (vladskakun) [reply tests/test_issue.py]: fixed

File: tests/test_issue.py (hunk 13 above)
Comment (vladskakun) [reply tests/test_issue.py]: fixed

File: tests/test_issue.py (hunk 15 above)
Comment (vladskakun) [reply tests/test_issue.py]: fixed

File: pages/Issues.py (hunk 16 above)
Comment (vladskakun) [reply pages/Issues.py]: fixed

File: pages/NavigationPage.py (hunk 17 above)
Comment (vladskakun) [reply pages/NavigationPage.py]: fixed

File: config.py [hunk 18]
```diff
@@ -0,0 +1,7 @@
+"""
+    Test data for logging in us admin
+"""
+admin_email = "admin@gmail.com"
+admin_pass = "1qaz1qaz"
+landing_sign_in_button = "Sign In"
```
Comment (TanyaKalistruk) [inline config.py:6]: Vote for using such constants in the top of related test file

File: pages/ModalPage.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        element.click()
+
+    def login(self, username, password):
+        """
+        Sign in method as a scenario
+        actor - is a person with own permissions (admin or user).
+        Method of precondition for testing
+        :param username: username
+        :param password: pass phrase
+        :return: login the EventExpress app
```
Comment (TanyaKalistruk) [inline pages/ModalPage.py:52]: But we don't have return in this method
Vote for removing this line

File: pages/common/BasePage.py [hunk 20]
```diff
@@ -0,0 +1,17 @@
+from pages.Issues import IssuesPage
+from pages.Landing import LandingPage
+from pages.ModalPage import ModalPage
+from pages.NavigationPage import NavigationPage
+from pages.common.baseWrapper import BaseWrapper
+
+
+class BasePage(BaseWrapper):
+    def __init__(self, driver):
```
Comment (TanyaKalistruk) [inline pages/common/BasePage.py:9]: I think that empty line is missed here
Applicable to the next file

File: pages/common/baseWrapper.py [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+from selenium.webdriver.support import expected_conditions as EC
+
+
+class BaseWrapper:
+    def __init__(self, driver):
+        """
+            Method for class fields declaration.
+        """
+        self.driver = driver
+        self.base_url = "https://eventsexpress-test.azurewebsites.net/"
```
Comment (TanyaKalistruk) [inline pages/common/baseWrapper.py:12]: We discussed that this should be moved to the constants

File: pages/common/baseWrapper.py [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+    def find_element_by_css(self, locator, time=10):
```
Comment (TanyaKalistruk) [inline pages/common/baseWrapper.py:14]: Probably timeout instead of time will be correctly

File: tests/conftest.py [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+    yield BasePage(driver)
+    driver.quit()
+
+
+def login_admin():
+    """
+        Login as an admin
+    """
+    with allure.step('Login as an admin'):
+        app.modal.login(config.admin_email, config.admin_pass)
```
Comment (TanyaKalistruk) [inline tests/conftest.py:20]:
```suggestion
@pytest.fixture(scope="function")
def admin_login(app):
    """
        Login as an admin
    """
        app.modal.login(config.admin_email, config.admin_pass)
```

File: config.py (hunk 18 above)
Comment (vladskakun) [reply config.py:6]: Do you mean before docstring?

File: tests/test_issue.py [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+    """
+    expected_result = 2
+    app.landing.go_to_site()
+    app.landing.sign_up_btn.click_btn_by_css()
+    app.modal.login(config.admin_email, config.admin_pass)
+    app.landing.find_event_btn.click_btn_by_css()
+    app.navigation.go_to_page("Issues")
+    app.issues.date_from_dtp.write_date_to_datepicker(21, 6, 2021)
+    app.issues.date_to_dtp.write_date_to_datepicker(15, 9, 2022)
+    app.issues.status_btns.click_btn_by_name("SEARCH")
```
[the remark this answers is not in the document]
Comment (vladskakun) [reply tests/test_issue.py:19]: Do you mean, to make these strings constants in this file or in the config file?

File: pages/common/baseWrapper.py (hunk 21 above)
Comment (vladskakun) [reply pages/common/baseWrapper.py]: done

File: pages/common/BasePage.py (hunk 20 above)
Comment (vladskakun) [reply pages/common/BasePage.py]: done

File: pages/ModalPage.py (hunk 19 above)
Comment (vladskakun) [reply pages/ModalPage.py]: done

File: config.py (hunk 18 above)
Comment (vladskakun) [reply config.py]: done

File: tests/test_issue.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (vladskakun) [reply tests/test_issue.py]: done

**Review by TanyaKalistruk — APPROVED** [review]

**Pull request merged**